src/analysis/normed_space/lp_space.lean

```diff
@@ -0,0 +1,516 @@
+/-
+Copyright (c) 2021 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import analysis.mean_inequalities
+import analysis.mean_inequalities_pow
+import topology.algebra.ordered.liminf_limsup
+
+/-!
+# ℓp space
+
+This file describes properties of elements `f` of a pi-type `Π i, E i` with finite "norm",
+defined for `p:ℝ≥0∞` as the size of the support of `f` if `p=0`, `(∑' a, ∥f a∥^p) ^ (1/p)` for
+`0 < p < ∞` and `⨆ a, ∥f a∥` for `p=∞`.
+
+The Prop-valued `mem_ℓp f p` states that a function `f : Π i, E i` has finite norm according
+to the above definition; that is, `f` has finite support if `p = 0`, `summable (λ a, ∥f a∥^p)` if
+`0 < p < ∞`, and `bdd_above (norm '' (set.range f))` if `p = ∞`.
+
+The space `lp E p` is the subtype of elements of `Π i : α, E i` which satisfy `mem_ℓp f p`. For
+`1 ≤ p`, the "norm" is genuinely a norm and (TODO) `lp` is a complete metric space.
+
+## Main definitions
+
+* `mem_ℓp f p` : property that the function `f` satisfies, as appropriate, `f` finitely supported
+  if `p = 0`, `summable (λ a, ∥f a∥^p)` if `0 < p < ∞`, and `bdd_above (norm '' (set.range f))` if
+  `p = ∞`
+* `lp E p` : elements of `Π i : α, E i` such that `mem_ℓp f p`. Defined as an `add_subgroup` of
+  `Π i : α, E i`.
+
+## Implementation
+
+Since `lp` is defined as an `add_subgroup`, dot notation does not work. Use `lp.norm_neg f` to
+say that `∥-f∥ = ∥f∥`, instead of the non-working `f.norm_neg`.
+
+## TODO
+
+* Hölder's inequality
+* Completeness of `lp`
+* Equivalence with `pi_Lp`, for `α` finite
+* Equivalence with `measure_theory.Lp`, for `f : α → E` (i.e., functions rather than pi-types) and
+  the counting measure on `α`
+* Equivalence with `bounded_continuous_function`, for `f : α → E` (i.e., functions rather than
+  pi-types) and `p = ∞`, and the discrete topology on `α`
+
+-/
+
+noncomputable theory
+open_locale nnreal ennreal big_operators
+
+variables {α : Type*} {E : α → Type*} {p q : ℝ≥0∞} [Π i, normed_group (E i)]
+
+/-!
+### `mem_ℓp` predicate
+
+-/
+
+/-- The property that `f : Π i : α, E i`
+* is finitely supported, if `p = 0`, or
+* admits an upper bound for `set.range (λ i, ∥f i∥)`, if `p = ∞`, or
+* has the series `∑' i, ∥f i∥ ^ p` be summable, if `0 < p < ∞`. -/
+def mem_ℓp (f : Π i, E i) (p : ℝ≥0∞) : Prop :=
```

Comment:
How about swapping the order of these arguments?
```suggestion
def mem_ℓp (p : ℝ≥0∞) (f : Π i, E i) : Prop :=
```
I think we generally care about `p` before we care about `f`.

Reply:
This way is consistent with `measure_theory.Lp` (but, on the other hand, your way is consistent with `pi_Lp`!)

Reply:
Pick whichever you prefer, it's not a big deal. Swapping arguments does allow you to write `carrier := mem_ℓp p` in the definition of `lp` if that tempts you ;-)

Reply:
I'll keep as is, I think, and maybe later we can select a standard order for all 3 versions.